### PR TITLE
Remove the Generic instance for LicenseId

### PR DESCRIPTION
**Pull request by Kleidukos**
This was one of the main culprits when analysing the resources taken
by GHC when compiling lib:Cabal.

Is related to #8074.

---
This is what the timings look with, and without the instance:

image

then 

45htwie6

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by Mikolaj** [comment]
Why is Setup trying to make up for the LicenseId speedup?

**Comment by Kleidukos** [comment]
Good question, this is indeed quite puzzling.

**Review by gbaz — APPROVED** [review]

**Comment by andreasabel** [comment]
Upstream (GHC) issue: https://gitlab.haskell.org/ghc/ghc/-/issues/16577

File: Cabal-tests/tests/UnitTests/Distribution/SPDX.hs [hunk 1]
```diff
@@ -25,15 +25,7 @@ import Test.QuickCheck.Instances.Cabal ()
 spdxTests :: [TestTree]
 spdxTests =
     [ testProperty "LicenseId roundtrip" licenseIdRoundtrip
-#if MIN_VERSION_binary(0,7,0)
```
Comment (andreasabel) [inline Cabal-tests/tests/UnitTests/Distribution/SPDX.hs:28]: These deleted test could be rather deactivated by placing them under
```c
#if false && MIN_VERSION_binary(0,7,0)
```
with a comment that they should be resurrected once the GHC is fixed and `Generic` instance of `LicenseId` is restored.

Comment (Kleidukos) [reply Cabal-tests/tests/UnitTests/Distribution/SPDX.hs:28]: @andreasabel According to every conversation I've had with GHC developers, including @bgamari, this is not a scenario that will arise before the heat death of the universe. :p

Comment (andreasabel) [reply Cabal-tests/tests/UnitTests/Distribution/SPDX.hs:28]: Ok, well then maybe we can afford the entropy of removing these tests altogether...

**Comment by andreasabel** [comment]
@Kleidukos @gbaz: I just had another idea:  If a large enum-type is problematic for GHC's deriving mechanism, how about breaking it up into smaller bits, and restore the original interface using `PatternSynonyms`?
This way maybe we can have the cake and eat it, too.

**Comment by Kleidukos** [comment]
@andreasabel It's an interesting solution, could you show us some benchmarks? I used https://github.com/codedownio/time-ghc-modules

**Comment by gbaz** [comment]
i think that idea is a lot of work for an instance that nobody really seems to need that much :-)

**Comment by andreasabel** [comment]
> @andreasabel It's an interesting solution, could you show us some benchmarks?

Na.  It is an idea, but I don't have the resources to follow up on it.

**Review by Mikolaj — APPROVED** [review]
LGTM. Let's merge.

**Pull request merged**